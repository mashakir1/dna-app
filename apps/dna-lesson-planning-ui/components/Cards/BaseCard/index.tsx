// Imports
import { ReactNode } from "react";

// Styles
import styles from "./BaseCard.module.scss";

interface BaseCardProps {
  bordered?: boolean;
  children: ReactNode;
  classNames?: string
  onClick?: () => void
}

export const BaseCard: (props: BaseCardProps) => JSX.Element = ({
  bordered, children, classNames, onClick
}) => (
  <div className={`${
    bordered ? styles.bordered : ""
  } ${
    styles.card
  } ${
    classNames ? classNames : ""
  }`} onClick={onClick}>{children}</div>
);
