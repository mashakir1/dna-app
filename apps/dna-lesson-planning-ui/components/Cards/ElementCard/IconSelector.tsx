// Components
import { CloseElementIcon, EmptyTick } from "components/Svg";
import { ClickableIcon } from "components/Icons";
import { useMemo } from "react";

interface IconSelectorProps {
  hovering: boolean;
  selected: boolean;
  clickable: boolean;
  stickyCross: boolean;
  crossOnClick?: () => void;
}

export const IconSelector: (props: IconSelectorProps) => JSX.Element = ({
  hovering, selected, clickable,
  stickyCross, crossOnClick
}) => {
  const defaultColour = "#ffffff33";
  const selectedColour = "#177584";

  const shouldShowCross = useMemo(
    () => stickyCross || (hovering && selected),
    [stickyCross, hovering, selected]
  );
  const shouldShowTick = useMemo(
    () => clickable && (hovering || selected),
    [clickable, hovering, selected]
  );

  if (shouldShowCross) {
    const closeIcon = <CloseElementIcon colour="white" fill={selected ? selectedColour : defaultColour} />;
    return crossOnClick
      ? <ClickableIcon icon={closeIcon} onClick={(event) => {
        event.stopPropagation();
        crossOnClick();
      }} />
      : closeIcon;
  }

  if (shouldShowTick) {
    return <EmptyTick colour="white" fill={selected ? selectedColour : defaultColour} />;
  }

  return <svg />;
};
