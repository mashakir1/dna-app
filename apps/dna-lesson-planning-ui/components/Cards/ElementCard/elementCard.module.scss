.elementCard {
  height: 130px;
  min-width: 130px;
  max-width: 130px;

  &.clickable {
    cursor: pointer;

    &:hover {
      .elementCardBody {
        background-color: $colour_background_grey;
      }
    }
  }

  & svg {
    height: 20px;
    width: 20px;
  }

  // TODO: sort out text wrapping on cards
  .elementCardBody {
    overflow: hidden;
    white-space: pre-wrap;
    text-overflow: ellipsis;
  }
}

.elementCardHeader {
  border-radius: 6px 6px 0 0;
  padding: $spacing_xs $spacing-ks;
  background-color: $colour_ui_teal;

  & p {
    line-height: 0;

    & > b {
      font-size: $type_size_xs;
      font-family: Montserrat-SemiBold;
      font-weight: 600;
      line-height: 12px;
      color: $colour_white;
    }
  }
}

.elementCardBody {
  height: 100%;
  padding: $spacing_ks;
  background-color: $colour_white;
  border: $colour_line_grey solid 1px;
  border-top: none;
  border-radius: 0 0 6px 6px;

  & p {
    font-size: $type_size_s;
    line-height: 16px;
  }

  & svg:hover path {
    fill: $colour_dark_text
  }
}

.elementCardFooter {
  p {
    line-height: 20px;
  }
}

.duration {
  color: adjust-color($colour_body_copy, $alpha: -0.5);
  text-transform: none;
}
