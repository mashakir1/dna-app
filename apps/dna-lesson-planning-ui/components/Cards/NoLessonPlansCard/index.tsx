// Imports
import useTranslation from "next-translate/useTranslation";
import Link from "next/link";
import Trans from "next-translate/Trans";

// Components
import { BaseCard } from "../BaseCard";
import { IconWithBackground } from "components/Icons";
import { FileIcon } from "components/Svg";

// Styles
import styles from "styles/pages/home.module.scss";

export const NoLessonPlansCard = () => {
  const { t } = useTranslation("common");

  return <BaseCard classNames={`${styles.emptyCard} flex column items-center gap-m`}>
    <IconWithBackground icon={<FileIcon />} />
    <p>{t("no-lesson-plans-to-show")}</p>
    <p><Trans
      i18nKey={"need-help"}
      components={[<Link href={"/"} key={""} />]}
      ns={"common"}
    /></p>
  </BaseCard>;
};
