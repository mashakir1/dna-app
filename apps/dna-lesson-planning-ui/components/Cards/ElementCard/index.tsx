// Imports
import { useCallback, useMemo, useState } from "react";

// Components
import { InfoIcon } from "components/Svg";
import { ClickableIcon } from "components/Icons";
import { IconSelector } from "./IconSelector";

// Styles
import styles from "./elementCard.module.scss";

/*
* How to use the ElementCard:
*  title, text, duration & subjectColour all show up as content on the card
*  selected determines if the card has been chosen by a user
*  onClick determines what happens when a user clicks on the card.
*    it also applies hover states
*  stickyCross will keep the crossIcon shown at all times
*  crossOnClick determines what will happen when the cross is clicked
*  infoOnClick determines what will happen when the infoIcon is clicked
*/

export interface ElementCardProps {
  id: string;
  title: string;
  text: string;
  duration: number;
  subjectColour: string;
  selected?: boolean;
  onClick?: (id: string) => void;
  stickyCross?: boolean;
  crossOnClick?: (id: string) => void;
  infoOnClick?: () => void;
}

export const ElementCard: (props: ElementCardProps) => JSX.Element = ({
  id, title, text, duration, subjectColour,
  selected, onClick, stickyCross, crossOnClick, infoOnClick
}) => {
  const clickable = useMemo(() => !!onClick, [onClick]);
  const [hovering, setHovering] = useState<boolean>(false);

  const onClickHandler = useCallback(() => onClick(id), [id, onClick]);
  const crossOnClickHandler = useCallback(() => crossOnClick(id), [id, crossOnClick]);

  return <div
    className={`${styles.elementCard} ${clickable ? styles.clickable : ""} flex column`}
    {...(onClick && { onClick: onClickHandler })}
    onMouseEnter={() => setHovering(true)}
    onMouseLeave={() => setHovering(false)}
  >
    <ElementCardHeader
      title={title}
      subjectColour={subjectColour}
      hovering={hovering}
      selected={selected}
      clickable={clickable}
      stickyCross={stickyCross}
      {...(crossOnClick && { crossOnClick: crossOnClickHandler })}
    />
    <ElementCardBody
      text={text}
      duration={duration}
      infoOnClick={infoOnClick}
    />
  </div>;
};

interface ElementCardHeaderProps {
  title: string;
  subjectColour: string;
  hovering: boolean;
  selected: boolean;
  clickable: boolean;
  stickyCross?: boolean;
  crossOnClick?: () => void;
}

const ElementCardHeader: (props: ElementCardHeaderProps) => JSX.Element = ({
  title, subjectColour, hovering,
  selected, clickable, stickyCross, crossOnClick
}) => {
  const headerColour = selected ? "#30CCD6" : subjectColour;

  return <div
    className={`${styles.elementCardHeader} flex justify-between items-center`}
    style={{ backgroundColor: headerColour }}
  >
    <p><b>{title}</b></p>
    <IconSelector
      hovering={hovering}
      selected={selected}
      clickable={clickable}
      stickyCross={stickyCross}
      crossOnClick={crossOnClick}
    />
  </div>;
};

interface ElementCardBodyProps {
  text: string;
  duration: number;
  infoOnClick: () => void;
}

const ElementCardBody: (props: ElementCardBodyProps) => JSX.Element = ({
  text, duration, infoOnClick
}) => {
  const handleInfoIconClick = (event) => {
    event.stopPropagation();
    infoOnClick();
  };

  return <div className={`${styles.elementCardBody} flex column justify-between`}>
    <p>{text}</p>
    <div className={`${styles.elementCardFooter} flex justify-between items-center`}>
      <p className={`${styles.duration} button small`}><b>{duration} mins</b></p>
      {infoOnClick && <ClickableIcon icon={<InfoIcon />} onClick={handleInfoIconClick} />}
    </div>
  </div>;
};
