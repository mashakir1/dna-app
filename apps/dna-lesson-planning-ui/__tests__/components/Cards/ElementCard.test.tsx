import { fireEvent, render, screen } from "@testing-library/react";
import { ElementCard, ElementCardProps } from "components/Cards/ElementCard";

jest.mock("components/Svg", () => ({
  EmptyTick() {
    return <div data-testid="emptyTick" />;
  },
  CloseElementIcon() {
    return <div data-testid="closeElement" />;
  },
  InfoIcon() {
    return <div data-testid="infoIcon" />;
  }
}));

describe("ElementCard", () => {
  const defaultProps: ElementCardProps = {
    id: "testId",
    title: "testTitle",
    text: "testText",
    duration: 0,
    subjectColour: "#ffffff33"
  };

  it("should render a basic ElementCard", () => {
    const { container } = render(<ElementCard {...defaultProps} />);

    expect(container).toMatchSnapshot();
    expect(screen.queryByText("testTitle")).toBeTruthy();
    expect(screen.queryByText("testText")).toBeTruthy();
  });

  it("should change headerColour when selected", () => {
    const { container } = render(<ElementCard {...defaultProps} selected />);

    expect(container).toMatchSnapshot();
  });

  it("should trigger onClick when clicked", () => {
    const testOnClick = jest.fn();
    const { container } = render(
      <ElementCard {...defaultProps} onClick={testOnClick} />
    );

    expect(container).toMatchSnapshot();
    fireEvent.click(screen.queryByText("testTitle"));
    expect(testOnClick).toHaveBeenCalled();
  });

  it("should show tick if hover and has an onClick", () => {
    const testOnClick = jest.fn();
    const { container } = render(
      <ElementCard {...defaultProps} onClick={testOnClick} />
    );

    expect(container).toMatchSnapshot();
    expect(screen.queryByTestId("emptyTick")).toBeFalsy();

    fireEvent.mouseEnter(screen.queryByText("testTitle"));
    expect(screen.queryByTestId("emptyTick")).toBeTruthy();

    fireEvent.mouseLeave(screen.queryByText("testTitle"));
    expect(screen.queryByTestId("emptyTick")).toBeFalsy();
  });

  it("should show tick if selected and has an onClick", () => {
    const testOnClick = jest.fn();
    const { container } = render(
      <ElementCard{...defaultProps} selected onClick={testOnClick} />
    );

    expect(container).toMatchSnapshot();
    expect(screen.queryByTestId("emptyTick")).toBeTruthy();
  });

  it("should show cross if selected, hover and has an onClick", () => {
    const testOnClick = jest.fn();
    const { container } = render(
      <ElementCard {...defaultProps} selected onClick={testOnClick} />);

    expect(container).toMatchSnapshot();
    expect(screen.queryByTestId("emptyTick")).toBeTruthy();
    expect(screen.queryByTestId("closeElement")).toBeFalsy();

    fireEvent.mouseEnter(screen.queryByText("testTitle"));

    expect(screen.queryByTestId("emptyTick")).toBeFalsy();
    expect(screen.queryByTestId("closeElement")).toBeTruthy();

    fireEvent.mouseLeave(screen.queryByText("testTitle"));

    expect(screen.queryByTestId("emptyTick")).toBeTruthy();
    expect(screen.queryByTestId("closeElement")).toBeFalsy();
  });

  it("should show cross when stickyCross", () => {
    const { container } = render(<ElementCard {...defaultProps} stickyCross />);

    expect(container).toMatchSnapshot();
    expect(screen.queryByTestId("closeElement")).toBeTruthy();
  });

  it("should trigger crossOnClick when clicking cross", () => {
    const testCrossOnClick = jest.fn();
    const { container } = render(
      <ElementCard {...defaultProps} stickyCross crossOnClick={testCrossOnClick} />
    );

    expect(container).toMatchSnapshot();
    fireEvent.click(screen.queryByTestId("closeElement"));
    expect(testCrossOnClick).toHaveBeenCalled();
  });

  it("should only trigger crossOnClick when clicking cross while both onClicks are present", () => {
    const testOnClick = jest.fn();
    const testCrossOnClick = jest.fn();
    const { container } = render(<ElementCard
      {...defaultProps} onClick={testOnClick}
      stickyCross crossOnClick={testCrossOnClick}
    />);

    expect(container).toMatchSnapshot();
    fireEvent.click(screen.queryByTestId("closeElement"));
    expect(testOnClick).not.toHaveBeenCalled();
    expect(testCrossOnClick).toHaveBeenCalled();
  });

  it("should trigger infoOnClick when clicking info", () => {
    const testInfoOnClick = jest.fn();
    const { container } = render(
      <ElementCard {...defaultProps} infoOnClick={testInfoOnClick} />
    );

    expect(container).toMatchSnapshot();
    expect(screen.queryByTestId("infoIcon")).toBeTruthy();
    fireEvent.click(screen.queryByTestId("infoIcon"));
    expect(testInfoOnClick).toHaveBeenCalled();
  });

  it("should only trigger infoOnClick when clicking info while both onClicks are present", () => {
    const testOnClick = jest.fn();
    const testInfoOnClick = jest.fn();
    const { container } = render(<ElementCard
      {...defaultProps} onClick={testOnClick} stickyCross infoOnClick={testInfoOnClick}
    />);

    expect(container).toMatchSnapshot();
    fireEvent.click(screen.queryByTestId("infoIcon"));
    expect(testOnClick).not.toHaveBeenCalled();
    expect(testInfoOnClick).toHaveBeenCalled();
  });
});
