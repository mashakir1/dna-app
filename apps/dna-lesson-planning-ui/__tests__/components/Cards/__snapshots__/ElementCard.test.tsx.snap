// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ElementCard should change headerColour when selected 1`] = `
<div>
  <div
    class="elementCard  flex column"
  >
    <div
      class="elementCardHeader flex justify-between items-center"
      style="background-color: rgb(48, 204, 214);"
    >
      <p>
        <b>
          testTitle
        </b>
      </p>
      <svg />
    </div>
    <div
      class="elementCardBody flex column justify-between"
    >
      <p>
        testText
      </p>
      <div
        class="elementCardFooter flex justify-between items-center"
      >
        <p
          class="duration button small"
        >
          <b>
            0
             mins
          </b>
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`ElementCard should only trigger crossOnClick when clicking cross while both onClicks are present 1`] = `
<div>
  <div
    class="elementCard clickable flex column"
  >
    <div
      class="elementCardHeader flex justify-between items-center"
      style="background-color: rgba(255, 255, 255, 0.2);"
    >
      <p>
        <b>
          testTitle
        </b>
      </p>
      <div
        class="clickableIcon flex"
      >
        <div
          data-testid="closeElement"
        />
      </div>
    </div>
    <div
      class="elementCardBody flex column justify-between"
    >
      <p>
        testText
      </p>
      <div
        class="elementCardFooter flex justify-between items-center"
      >
        <p
          class="duration button small"
        >
          <b>
            0
             mins
          </b>
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`ElementCard should only trigger infoOnClick when clicking info while both onClicks are present 1`] = `
<div>
  <div
    class="elementCard clickable flex column"
  >
    <div
      class="elementCardHeader flex justify-between items-center"
      style="background-color: rgba(255, 255, 255, 0.2);"
    >
      <p>
        <b>
          testTitle
        </b>
      </p>
      <div
        data-testid="closeElement"
      />
    </div>
    <div
      class="elementCardBody flex column justify-between"
    >
      <p>
        testText
      </p>
      <div
        class="elementCardFooter flex justify-between items-center"
      >
        <p
          class="duration button small"
        >
          <b>
            0
             mins
          </b>
        </p>
        <div
          class="clickableIcon flex"
        >
          <div
            data-testid="infoIcon"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ElementCard should render a basic ElementCard 1`] = `
<div>
  <div
    class="elementCard  flex column"
  >
    <div
      class="elementCardHeader flex justify-between items-center"
      style="background-color: rgba(255, 255, 255, 0.2);"
    >
      <p>
        <b>
          testTitle
        </b>
      </p>
      <svg />
    </div>
    <div
      class="elementCardBody flex column justify-between"
    >
      <p>
        testText
      </p>
      <div
        class="elementCardFooter flex justify-between items-center"
      >
        <p
          class="duration button small"
        >
          <b>
            0
             mins
          </b>
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`ElementCard should show cross if selected, hover and has an onClick 1`] = `
<div>
  <div
    class="elementCard clickable flex column"
  >
    <div
      class="elementCardHeader flex justify-between items-center"
      style="background-color: rgb(48, 204, 214);"
    >
      <p>
        <b>
          testTitle
        </b>
      </p>
      <div
        data-testid="emptyTick"
      />
    </div>
    <div
      class="elementCardBody flex column justify-between"
    >
      <p>
        testText
      </p>
      <div
        class="elementCardFooter flex justify-between items-center"
      >
        <p
          class="duration button small"
        >
          <b>
            0
             mins
          </b>
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`ElementCard should show cross when stickyCross 1`] = `
<div>
  <div
    class="elementCard  flex column"
  >
    <div
      class="elementCardHeader flex justify-between items-center"
      style="background-color: rgba(255, 255, 255, 0.2);"
    >
      <p>
        <b>
          testTitle
        </b>
      </p>
      <div
        data-testid="closeElement"
      />
    </div>
    <div
      class="elementCardBody flex column justify-between"
    >
      <p>
        testText
      </p>
      <div
        class="elementCardFooter flex justify-between items-center"
      >
        <p
          class="duration button small"
        >
          <b>
            0
             mins
          </b>
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`ElementCard should show tick if hover and has an onClick 1`] = `
<div>
  <div
    class="elementCard clickable flex column"
  >
    <div
      class="elementCardHeader flex justify-between items-center"
      style="background-color: rgba(255, 255, 255, 0.2);"
    >
      <p>
        <b>
          testTitle
        </b>
      </p>
      <svg />
    </div>
    <div
      class="elementCardBody flex column justify-between"
    >
      <p>
        testText
      </p>
      <div
        class="elementCardFooter flex justify-between items-center"
      >
        <p
          class="duration button small"
        >
          <b>
            0
             mins
          </b>
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`ElementCard should show tick if selected and has an onClick 1`] = `
<div>
  <div
    class="elementCard clickable flex column"
  >
    <div
      class="elementCardHeader flex justify-between items-center"
      style="background-color: rgb(48, 204, 214);"
    >
      <p>
        <b>
          testTitle
        </b>
      </p>
      <div
        data-testid="emptyTick"
      />
    </div>
    <div
      class="elementCardBody flex column justify-between"
    >
      <p>
        testText
      </p>
      <div
        class="elementCardFooter flex justify-between items-center"
      >
        <p
          class="duration button small"
        >
          <b>
            0
             mins
          </b>
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`ElementCard should trigger crossOnClick when clicking cross 1`] = `
<div>
  <div
    class="elementCard  flex column"
  >
    <div
      class="elementCardHeader flex justify-between items-center"
      style="background-color: rgba(255, 255, 255, 0.2);"
    >
      <p>
        <b>
          testTitle
        </b>
      </p>
      <div
        class="clickableIcon flex"
      >
        <div
          data-testid="closeElement"
        />
      </div>
    </div>
    <div
      class="elementCardBody flex column justify-between"
    >
      <p>
        testText
      </p>
      <div
        class="elementCardFooter flex justify-between items-center"
      >
        <p
          class="duration button small"
        >
          <b>
            0
             mins
          </b>
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`ElementCard should trigger infoOnClick when clicking info 1`] = `
<div>
  <div
    class="elementCard  flex column"
  >
    <div
      class="elementCardHeader flex justify-between items-center"
      style="background-color: rgba(255, 255, 255, 0.2);"
    >
      <p>
        <b>
          testTitle
        </b>
      </p>
      <svg />
    </div>
    <div
      class="elementCardBody flex column justify-between"
    >
      <p>
        testText
      </p>
      <div
        class="elementCardFooter flex justify-between items-center"
      >
        <p
          class="duration button small"
        >
          <b>
            0
             mins
          </b>
        </p>
        <div
          class="clickableIcon flex"
        >
          <div
            data-testid="infoIcon"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ElementCard should trigger onClick when clicked 1`] = `
<div>
  <div
    class="elementCard clickable flex column"
  >
    <div
      class="elementCardHeader flex justify-between items-center"
      style="background-color: rgba(255, 255, 255, 0.2);"
    >
      <p>
        <b>
          testTitle
        </b>
      </p>
      <svg />
    </div>
    <div
      class="elementCardBody flex column justify-between"
    >
      <p>
        testText
      </p>
      <div
        class="elementCardFooter flex justify-between items-center"
      >
        <p
          class="duration button small"
        >
          <b>
            0
             mins
          </b>
        </p>
      </div>
    </div>
  </div>
</div>
`;
